File: packages/prompts/loader.py
from __future__ import annotations

import json
import re
from pathlib import Path

SRC_DIR = Path(__file__).parent / "src"


def load_prompt(
    name: str,
    lang: str,
    variables: dict[str, str],
    *,
    version: str | None = None,
) -> str:
    prompt_dir = SRC_DIR / name
    meta_path = prompt_dir / "meta.json"
    meta = json.loads(meta_path.read_text(encoding="utf-8"))

    resolved_version = version or meta["current"]
    file_path = prompt_dir / f"{resolved_version}.{lang}.md"

    if not file_path.exists():
        msg = f"Prompt file not found: {file_path}"
        raise FileNotFoundError(msg)

    template = file_path.read_text(encoding="utf-8")

    missing: list[str] = []

    def _replace(match: re.Match[str]) -> str:
        key = match.group(1)
        if key in variables:
            return variables[key]
        missing.append(key)
        return match.group(0)

    result = re.sub(r"\{\{(\w+)\}\}", _replace, template)

    if missing:
        msg = f"Missing prompt variables: {', '.join(missing)}"
        raise ValueError(msg)

    return result

File: packages/prompts/loader.ts
import * as fs from 'node:fs';
import * as path from 'node:path';

type Meta = {
  current: string;
  frozen: string[];
};

const SRC_DIR = path.join(__dirname, 'src');

export function loadPrompt(
  name: string,
  lang: string,
  variables: Record<string, string>,
  version?: string,
): string {
  const promptDir = path.join(SRC_DIR, name);
  const metaPath = path.join(promptDir, 'meta.json');
  const meta: Meta = JSON.parse(fs.readFileSync(metaPath, 'utf-8'));

  const resolvedVersion = version ?? meta.current;
  const filePath = path.join(promptDir, `${resolvedVersion}.${lang}.md`);

  if (!fs.existsSync(filePath)) {
    throw new Error(`Prompt file not found: ${filePath}`);
  }

  let template = fs.readFileSync(filePath, 'utf-8');

  const missing: string[] = [];
  template = template.replace(/\{\{(\w+)\}\}/g, (match, key: string) => {
    if (key in variables) {
      return variables[key];
    }
    missing.push(key);
    return match;
  });

  if (missing.length > 0) {
    throw new Error(`Missing prompt variables: ${missing.join(', ')}`);
  }

  return template;
}
